src/index/bip352.cpp

```diff
@@ -84,15 +84,21 @@ bool BIP352Index::GetSilentPaymentKeys(std::vector<CTransactionRef> txs, CBlockU
                 LOCK(cs_main);
                 const CCoinsViewCache& coins_cache = m_chainstate->CoinsTip();
 
-                uint32_t spent{0};
+                uint32_t discarded{0};
                 for (uint32_t i{0}; i < tx->vout.size(); i++) {
                     COutPoint outpoint(tx->GetHash(), i);
+                    // Discard dust
+                    if (tx->vout.at(i).nValue < 1000) {
+                        discarded++;
+                        continue;
+                    }
+
                     // Many new blocks may be processed while generating the index,
                     // in between HaveCoin calls. This is not a problem, because
                     // the cut-through index can safely have false positives.
```

Comment:
Note that if we serve this via a BIP157 the cut-through version is going to be indeterministic. Same if people pick a different dust limit. This seems hard to avoid in any case, because the entire filter chain potentially changes with each new block that comes in.

Reply:
Yeah, with cut-through and a dust filter, this starts to feel like not a good fit for a BIP157 style filter, only because we'd basically have to get rid of the filter chain and hashes to avoid making the server constantly recalculate. Another thing about dust filters is we could make it client query param, i.e. the server still keeps everything and then the client says "only send me stuff where the UTXOs have a value of >= X". More storage space for the server, but clients then only download tweak data that they *actually* intend to spend. The nice thing about this model is the client can always query with a lower dust limit at any point to get everything else.

Got a few ideas on how we could make this work with a BIP157, but need to give it some more thought. It might also be case of trying to fit a square peg into a round hole, so maybe its better to start thinking from scratch on the protocol for serving the data.

(one obvious answer is we just keep the index in core and expose it via RPC, and then the exact p2p details are handled by something which uses core as a backend, e.g. electrumX).

Reply:
> only send me stuff where the UTXOs have a value of >= X

So now we need an index and efficient sorting.

> It might also be case of trying to fit a square peg into a round hole

For the index without cut-through I think BIP157 is fine. But to get the extra bandwidth savings of cut-through and a (custom) dust limit, it indeed seems we need something else. But ideally that shouldn't block shipping something that works, even if it uses a bit more bandwidth that we'd like.

It's worth considering at least adding the 546 sat dust limit to BIP352. Or at the very minimum exclude 0 sat outputs. I'll re-run the normal index to see the effect.

Reply:
> For the index without cut-through I think BIP157 is fine

Agree.

> It's worth considering at least adding the 546 sat dust limit to BIP352

I'd argue these are all implementation details/optimizations. The BIP itself should be reserved for describing the protocol. There also might be use cases where where the value of the output is not the primary concern.

Reply:
I just checked two things:

1. LevelDB does not support indexes. It also does not support SQL queries.

If we were to store the amount (of the largest taproot output) in the index, any value filtering has to be done at the time we query the index. Worst case 9000 entries per block (a 1 input 1 output taproot transaction is 111 vbyte). That might be fine, you'd have to benchmark.

If it _is_ a bottleneck, then we could refactor the base index to allow sqlite3 as a backend.

2. The only way to update an entry is to `Delete(key)` and `Put(key)`. I currently use block hash as the key, so the ordering wouldn't change. But what happens to the rest of the index file as you do this?

According to `leveldb/doc/impl.md`:

> When the size of level L exceeds its limit, we compact it in a background
thread. 

> Compactions drop overwritten values. They also drop deletion markers if there
are no higher numbered levels that contain a file whose range overlaps the
current key.

So if we (occasionally) override a block entry to update its cut-through, LevelDB should take care of defragging it ("compaction"). We should keep an eye on the resource usage there.

Reply:
Using `IsDust()` makes no different for the regular index: https://github.com/Sjors/bitcoin/commit/20aba5dfa34224432787ac450b8e9d3a66d8eea9

Reply:
> Filtering < 546 dust makes no difference on the regular index size (unless I did something wrong in [411f035](https://github.com/bitcoin/bitcoin/commit/411f035f0e09142311b6a829490f175582b07096))

That's interesting you could compare with [Blindbit Oracle](https://github.com/setavenger/blindbit-oracle). There is an exposed endpoint that filters based on a defined dust limit. There definitely is a difference at ~1000. I'm not sure whether I've tested <546 yet.

Reply:
Not super surprising since this is a standardness limit, no? I don't think you can broadcast transactions without the output amount being greater than the dust limit (unless it is an `OP_RETURN`).

Reply:
> Filtering < 546 dust

I should add that I was using `IsDust()` with `DUST_RELAY_TX_FEE`, which takes SegWit into account to decide the exact dust limit.

> 10% of taproot UTXOs

I looked at the largest output, not all outputs.

Reply:
> 
> I should add that I was using `IsDust()` with `DUST_RELAY_TX_FEE`, which takes SegWit into account to decide the exact dust limit.
> 

Ah I'm not sure how this should affect the results tbh. In my index I just looked at the UTXO value, no modifications.

> 
> I looked at the largest output, not all outputs.

Yeah but my current understanding is that in order to see no difference on the index size at all, all of those UTXOs must have been in combination with bigger UTXOs, not eligible at all or somehow affected by the `DUST_RELAY_TX_FEE` modification (not sure how that exactly works). 
How are the dust limits stored in the index? If they are based on the normal UTXO values we could do the same comparison as before https://github.com/bitcoin/bitcoin/pull/28241#issuecomment-2076895408

Reply:
Dust limits are not stored in the index. https://github.com/bitcoin/bitcoin/pull/28241/commits/835dd4ba1ca4cc1abfd338e6bc38e25a116f5f8b stores the maximum output amount for each transaction. You can then use the RPC with any dust limit value you like.

Reply:
My bad, that's what I meant. So you can just say give me everything where at least one UTXO exists above that threshold value. That's what my implementation does as well. Should be comparable then.

Reply:
The [tweaks endpoint](https://github.com/setavenger/blindbit-oracle?tab=readme-ov-file#endpoints) with the `dustLimit` parameter. Any value should work I believe.

Reply:
I'll try that with 992 once I rebuilt the database. Is `dustLimit` `>= 992` or `> 992`?

Reply:
Oh but I don't want to use cut-through; that can only be compared when I generate the index at exactly the same height.

Reply:
hmm that's true, maybe I can quickly hack something together. It's probably better to compare the base index anyways.

Reply:
Took a bit longer than I expected. I've pushed to the [dev branch](https://github.com/setavenger/blindbit-oracle/tree/dev). It seems to be working as intended. No cut-through and you can choose any value you like for a dustLimit. The request looks like this `"localhost:8000/tweak-index/840000?dustLimit=546"` (`tweak-index` then only returns tweaks where the highest value is `>= 546`)

Reply:
```
curl -s localhost:8000/tweak-index/709632?dustLimit=992 | jq
{
  "error": "Server does not allow dustLimits"
}
```

Do I need to add a setting or rebuild the index?

Reply:
Restarting with `tweaks_full_with_dust_filter = 1`

Reply:
Sorry, should have mentioned that before. Yes, that flag/setting is required.

Reply:
This branch should have a sorted index. We can now find the txid a lot easier, I think.

Reply:
I added a fix for https://github.com/bitcoin/bitcoin/pull/28241/files#diff-1b4444ba5939fff2ab9f70d1d0d29fc2653b09a332e462166e01891e261c9fa9 and rebuilding the index now.

Reply:
After my fix I found another inconsistency, at block [709,640](https://mempool.space/block/0000000000000000000a3a458944554e24d1ef8d3ef8707067b01dcaf71526cd?page=39) and dust limit 992 your index finds 3 while Bitcoin Core finds 2.

I manually looked up the eligible transactions:
1. https://mempool.space/tx/350f06dff749f910b0ad8d918496850c676c22854e24a046304e73bdfe399a02
2. https://mempool.space/tx/0fdda311049e8a1450ef08bab4076c0e00f8e8334c861ed441f24980b8ac6a90
3. https://mempool.space/tx/73d535f1384c6b43280dc1974b3bcace1940a54c2ea675bad386923ec1e95007

Transaction (1) has one P2TR output of 897, which should get filtered. The second output of 1500 shouldn't count, because it's not taproot.

So I think the bug is on your end this time @setavenger :-)

Reply:
Wow completely missed the messages here. Yeah, was able to recreate the issue. I'll figure out what the issue is and come back to you after I've got a fix for the problem.

Reply:
@Sjors it's fixed https://github.com/setavenger/blindbit-oracle/pull/14 now. There was no taproot filter on the biggest utxo selection.

Reply:
Thanks, trying again...

Reply:
Matches!

Reply:
Cool! Would you mind sharing the current disk usage here https://github.com/setavenger/blindbit-oracle/issues/15
it's been a while since I've done a full mainnet sync.